In [1]:
import pandas as pd
import numpy as np

In [2]:
filepath = ('/Users/shijizhao/Documents/UCI/Courses/'
            'Intro to Data Science/Final Project/'
            'csv_hpa_2012_2016/ss16hpa.csv')

In [3]:
df = pd.read_csv(filepath)

In [4]:
pd.set_option('display.max_columns', None)
df.head(20)

RT       SERIALNO  DIVISION  PUMA  REGION  ST   ADJHSG   ADJINC  WGTP  NP  \
0   H  2012000000003         2  2100       1  42  1045360  1056030    12   1   
1   H  2012000000006         2  3102       1  42  1045360  1056030    13   4   
2   H  2012000000007         2  3211       1  42  1045360  1056030    19   1   
3   H  2012000000053         2  2200       1  42  1045360  1056030    12   1   
4   H  2012000000075         2  2500       1  42  1045360  1056030    13   6   
5   H  2012000000079         2  3900       1  42  1045360  1056030     8   3   
6   H  2012000000096         2  3003       1  42  1045360  1056030    29   2   
7   H  2012000000144         2  3402       1  42  1045360  1056030    24   4   
8   H  2012000000192         2  2803       1  42  1045360  1056030     0   1   
9   H  2012000000221         2  3209       1  42  1045360  1056030    18   2   
10  H  2012000000279         2  2402       1  42  1045360  1056030    32   2   
11  H  2012000000288         2  3800       1  42  1045360  1056030     7   4   
12  H  2012000000316         2  3602       1  42  1045360  1056030    30   4   
13  H  2012000000337         2  2302       1  42  1045360  1056030     0   1   
14  H  2012000000427         2  3201       1  42  1045360  1056030     0   1   
15  H  2012000000452         2  3303       1  42  1045360  1056030    13   1   
16  H  2012000000540         2  1300       1  42  1045360  1056030     6   0   
17  H  2012000000572         2  3304       1  42  1045360  1056030     0   1   
18  H  2012000000578         2  2801       1  42  1045360  1056030     6   0   
19  H  2012000000615         2  3207       1  42  1045360  1056030    14   1   

    TYPE  ACR  AGS  BATH  BDSP  BLD  BUS   CONP   ELEP   FS    FULP   GASP  \
0      1  NaN  NaN   1.0   1.0  9.0  NaN    NaN   20.0  1.0     2.0    3.0   
1      1  1.0  NaN   1.0   3.0  3.0  2.0    0.0  100.0  2.0     2.0  130.0   
2      1  1.0  NaN   1.0   3.0  3.0  2.0    0.0   60.0  1.0     2.0  320.0   
3      1  1.0  NaN   1.0   4.0  2.0  2.0    0.0   20.0  2.0  1000.0    3.0   
4      1  1.0  NaN   1.0   3.0  1.0  2.0    0.0   70.0  1.0     2.0    3.0   
5      1  1.0  NaN   1.0   3.0  2.0  2.0    0.0   80.0  2.0  1800.0  140.0   
6      1  1.0  NaN   1.0   4.0  2.0  2.0    0.0  110.0  2.0  3100.0   30.0   
7      1  1.0  NaN   1.0   4.0  2.0  2.0    0.0  450.0  2.0     2.0  120.0   
8      2  NaN  NaN   NaN   NaN  NaN  NaN    NaN    NaN  2.0     NaN    NaN   
9      1  NaN  NaN   1.0   1.0  9.0  NaN  390.0   60.0  2.0     2.0    1.0   
10     1  2.0  1.0   1.0   3.0  2.0  2.0    0.0  160.0  2.0     2.0   70.0   
11     1  1.0  NaN   1.0   3.0  2.0  2.0    0.0  110.0  2.0  1200.0    3.0   
12     1  2.0  1.0   1.0   3.0  2.0  2.0    0.0  200.0  2.0     2.0    3.0   
13     2  NaN  NaN   NaN   NaN  NaN  NaN    NaN    NaN  1.0     NaN    NaN   
14     2  NaN  NaN   NaN   NaN  NaN  NaN    NaN    NaN  2.0     NaN    NaN   
15     1  1.0  NaN   1.0   3.0  3.0  2.0    0.0   70.0  2.0     2.0   30.0   
16     1  2.0  NaN   1.0   1.0  1.0  2.0    NaN    NaN  NaN     NaN    NaN   
17     2  NaN  NaN   NaN   NaN  NaN  NaN    NaN    NaN  2.0     NaN    NaN   
18     1  1.0  NaN   1.0   3.0  2.0  2.0    NaN    NaN  NaN     NaN    NaN   
19     1  NaN  NaN   1.0   1.0  5.0  NaN    NaN  100.0  1.0     2.0   60.0   

    HFL    INSP    MHP  MRGI    MRGP  MRGT  MRGX  REFR  RMSP  RNTM   RNTP  \
0   8.0     NaN    NaN   NaN     NaN   NaN   NaN   1.0   3.0   2.0   70.0   
1   1.0   400.0    NaN   1.0  1400.0   1.0   1.0   1.0   6.0   NaN    NaN   
2   1.0     0.0    NaN   NaN     NaN   NaN   3.0   1.0   6.0   NaN    NaN   
3   4.0   560.0    NaN   NaN     NaN   NaN   3.0   1.0   7.0   NaN    NaN   
4   3.0     0.0  200.0   NaN     NaN   NaN   3.0   1.0   7.0   NaN    NaN   
5   1.0   330.0    NaN   NaN     NaN   NaN   3.0   1.0   6.0   NaN    NaN   
6   4.0  1000.0    NaN   1.0  1800.0   1.0   1.0   1.0   8.0   NaN    NaN   
7   1.0  1500.0    NaN   2.0  4500.0   1.0   1.0   1.0  11.0   NaN 

In [5]:
df.columns

Index(['RT', 'SERIALNO', 'DIVISION', 'PUMA', 'REGION', 'ST', 'ADJHSG',
       'ADJINC', 'WGTP', 'NP',
       ...
       'WGTP71', 'WGTP72', 'WGTP73', 'WGTP74', 'WGTP75', 'WGTP76', 'WGTP77',
       'WGTP78', 'WGTP79', 'WGTP80'],
      dtype='object', length=208)

In [6]:
# Select columns
df_col_filtered = df.loc[:,['NP','ACR','BATH','BDSP','BLD','CONP','ELEP','FS','FULP','GASP','HFL',
                        'INSP','REFR','RMSP','RWAT','RWATPR','SINK','STOV','TEN','TOIL','VALP','VEH',
                        'WATP','YBL','FES','FPARC','HHL','HHT','HINCP','HUGCL','HUPAC','HUPAOC',
                         'HUPARC','KIT','LNGI','MULTG','MV','NOC','NPF','NPP','NR','NRC','PARTNER','PLM',
                        'PSF','R18','R60','R65','RESMODE','TAXP','WIF','WKEXREL','WORKSTAT']]

In [7]:
pd.set_option('display.max_columns', None)
df_col_filtered.head()

NP  ACR  BATH  BDSP  BLD  CONP   ELEP   FS    FULP   GASP  HFL   INSP  \
0   1  NaN   1.0   1.0  9.0   NaN   20.0  1.0     2.0    3.0  8.0    NaN   
1   4  1.0   1.0   3.0  3.0   0.0  100.0  2.0     2.0  130.0  1.0  400.0   
2   1  1.0   1.0   3.0  3.0   0.0   60.0  1.0     2.0  320.0  1.0    0.0   
3   1  1.0   1.0   4.0  2.0   0.0   20.0  2.0  1000.0    3.0  4.0  560.0   
4   6  1.0   1.0   3.0  1.0   0.0   70.0  1.0     2.0    3.0  3.0    0.0   

   REFR  RMSP  RWAT  RWATPR  SINK  STOV  TEN  TOIL      VALP  VEH   WATP  YBL  \
0   1.0   3.0   1.0     9.0   1.0   1.0  3.0   1.0       NaN  1.0    2.0  5.0   
1   1.0   6.0   1.0     9.0   1.0   1.0  1.0   1.0  235000.0  2.0  800.0  9.0   
2   1.0   6.0   1.0     9.0   1.0   1.0  2.0   1.0  179000.0  0.0  420.0  1.0   
3   1.0   7.0   1.0     9.0   1.0   1.0  2.0   1.0  120000.0  1.0  480.0  1.0   
4   1.0   7.0   1.0     9.0   1.0   1.0  2.0   1.0    6000.0  2.0  200.0  6.0   

   FES  FPARC  HHL  HHT    HINCP  HUGCL  HUPAC  HUPAOC  HUPARC  KIT  LNGI  \
0  NaN    NaN  1.0  4.0   4600.0    0.0    4.0     4.0     4.0  1.0   1.0   
1  1.0    3.0  1.0  1.0  71000.0    0.0    3.0     3.0     3.0  1.0   1.0   
2  NaN    NaN  1.0  6.0   5000.0    0.0    4.0     4.0     4.0  1.0   1.0   
3  NaN    NaN  1.0  6.0  17200.0    0.0    4.0     4.0     4.0  1.0   1.0   
4  1.0    1.0  1.0  1.0  52600.0    1.0    1.0     4.0     1.0  1.0   1.0   

   MULTG   MV  NOC  NPF  NPP   NR  NRC  PARTNER  PLM  PSF  R18  R60  R65  \
0    1.0  1.0  0.0  NaN  0.0  0.0  0.0      0.0  1.0  0.0  0.0  0.0  0.0   
1    1.0  4.0  2.0  4.0  0.0  0.0  2.0      0.0  1.0  0.0  1.0  0.0  0.0   
2    1.0  6.0  0.0  NaN  0.0  0.0  0.0      0.0  1.0  0.0  0.0  1.0  1.0   
3    1.0  7.0  0.0  NaN  0.0  0.0  0.0      0.0  1.0  0.0  0.0  1.0  1.0   
4    2.0  3.0  0.0  6.0  0.0  0.0  2.0      0.0  1.0  1.0  1.0  0.0  0.0   

   RESMODE  TAXP  WIF  WKEXREL  WORKSTAT  
0      1.0   NaN  NaN      NaN       NaN  
1      1.0  43.0  2.0      4.0       1.0  
2      1.0  16.0  NaN      NaN       NaN  
3      2.0  29.0  NaN      NaN       NaN  
4      1.0   2.0  3.0      4.0       1.0

In [8]:
df_col_filtered.shape

(317416, 53)

In [9]:
df_col_filtered.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 317416 entries, 0 to 317415
Data columns (total 53 columns):
NP          317416 non-null int64
ACR         237248 non-null float64
BATH        279431 non-null float64
BDSP        279431 non-null float64
BLD         279431 non-null float64
CONP        191909 non-null float64
ELEP        254398 non-null float64
FS          292383 non-null float64
FULP        254398 non-null float64
GASP        254398 non-null float64
HFL         254398 non-null float64
INSP        191909 non-null float64
REFR        279431 non-null float64
RMSP        279431 non-null float64
RWAT        279431 non-null float64
RWATPR      279431 non-null float64
SINK        279431 non-null float64
STOV        279431 non-null float64
TEN         254398 non-null float64
TOIL        279431 non-null float64
VALP        194338 non-null float64
VEH         254398 non-null float64
WATP        254398 non-null float64
YBL         279431 non-null float64
FES         168177 non-null

In [10]:
df_col_filtered['TEN'].unique()

array([ 3.,  1.,  2., nan,  4.])

In [11]:
# select rows based on TEN, the target data. The values have to be either
# 1.0: Owned with mortgage or loan (include home equity loans) or
# 2.0: Owed free and clear
df_col_row_filtered = df_col_filtered.loc[(df_col_filtered['TEN'] == 1.0) |
                                          (df_col_filtered['TEN'] == 2.0), :]

In [12]:
df_col_row_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 191909 entries, 1 to 317415
Data columns (total 53 columns):
NP          191909 non-null int64
ACR         187801 non-null float64
BATH        191909 non-null float64
BDSP        191909 non-null float64
BLD         191909 non-null float64
CONP        191909 non-null float64
ELEP        191909 non-null float64
FS          191909 non-null float64
FULP        191909 non-null float64
GASP        191909 non-null float64
HFL         191909 non-null float64
INSP        191909 non-null float64
REFR        191909 non-null float64
RMSP        191909 non-null float64
RWAT        191909 non-null float64
RWATPR      191909 non-null float64
SINK        191909 non-null float64
STOV        191909 non-null float64
TEN         191909 non-null float64
TOIL        191909 non-null float64
VALP        191909 non-null float64
VEH         191909 non-null float64
WATP        191909 non-null float64
YBL         191909 non-null float64
FES         140070 non-null

Based on the non-null information above, there are 4 non-null numbers:
1. 191909: All data points are non-null;
2. 187801: NaN should be mobile house / not a one-family house;
3. 140539: NaN should be non-family household;
4. 140070: NaN should include non-family household / same-sex married couple;

In [13]:
# Firstly, for those variables affected by "non-family household"
# ['FPARC','NPF','WIF','WKEXREL','FES','WORKSTAT'], set 
# NaN as -1.

for var in ['FPARC','NPF','WIF','WKEXREL','FES','WORKSTAT']:
    df_col_row_filtered[var].fillna(-1,inplace=True)

# However, variables['FES','WORKSTAT'] are also affected
# by "same-sex married couple", set those values as 0.

df_col_row_filtered.loc[(df_col_row_filtered['FPARC']!=-1)&
                        (df_col_row_filtered['FES']==-1),['FES','WORKSTAT']] = 0

# Finally, the variable 'ACR' is affected by mobile house 
# / not a one-family house, set NaN as 0
df_col_row_filtered['ACR'].fillna(0,inplace=True)

/Users/shijizhao/anaconda3/lib/python3.6/site-packages/pandas/core/generic.py:6130: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)
/Users/shijizhao/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [14]:
df_col_row_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 191909 entries, 1 to 317415
Data columns (total 53 columns):
NP          191909 non-null int64
ACR         191909 non-null float64
BATH        191909 non-null float64
BDSP        191909 non-null float64
BLD         191909 non-null float64
CONP        191909 non-null float64
ELEP        191909 non-null float64
FS          191909 non-null float64
FULP        191909 non-null float64
GASP        191909 non-null float64
HFL         191909 non-null float64
INSP        191909 non-null float64
REFR        191909 non-null float64
RMSP        191909 non-null float64
RWAT        191909 non-null float64
RWATPR      191909 non-null float64
SINK        191909 non-null float64
STOV        191909 non-null float64
TEN         191909 non-null float64
TOIL        191909 non-null float64
VALP        191909 non-null float64
VEH         191909 non-null float64
WATP        191909 non-null float64
YBL         191909 non-null float64
FES         191909 non-null

In [15]:
# Change dtype of non-numerical variables
#for var in ['ACR','BATH','BLD','FS','HFL',
#            'REFR','RWAT','RWATPR','SINK','STOV','TEN','TOIL','VEH',
#            'YBL','FES','FPARC','HHL','HHT','HUGCL','HUPAC','HUPAOC',
#            'HUPARC','KIT','LNGI','MULTG','MV','NPP','NR','PARTNER','PLM',
#            'PSF','R18','R60','R65','RESMODE','TAXP','WIF','WKEXREL','WORKSTAT']:
#    df_col_row_filtered[var]=df_col_row_filtered[var].astype('category')
    
#for var in ['BDSP','CONP','ELEP','FULP','GASP','INSP','RMSP',
#           'VALP','WATP','FINCP','HINCP','NOC','NPF','NRC',]:
#    df_col_filtered[var]=df_col_filtered[var].astype('int')

In [16]:
df_col_row_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 191909 entries, 1 to 317415
Data columns (total 53 columns):
NP          191909 non-null int64
ACR         191909 non-null float64
BATH        191909 non-null float64
BDSP        191909 non-null float64
BLD         191909 non-null float64
CONP        191909 non-null float64
ELEP        191909 non-null float64
FS          191909 non-null float64
FULP        191909 non-null float64
GASP        191909 non-null float64
HFL         191909 non-null float64
INSP        191909 non-null float64
REFR        191909 non-null float64
RMSP        191909 non-null float64
RWAT        191909 non-null float64
RWATPR      191909 non-null float64
SINK        191909 non-null float64
STOV        191909 non-null float64
TEN         191909 non-null float64
TOIL        191909 non-null float64
VALP        191909 non-null float64
VEH         191909 non-null float64
WATP        191909 non-null float64
YBL         191909 non-null float64
FES         191909 non-null

In [17]:
df_col_row_filtered['FPARC'].value_counts()

 4.0    91240
-1.0    51370
 2.0    33526
 1.0     8086
 3.0     7687
Name: FPARC, dtype: int64

In [18]:
df_col_row_filtered['FES'].value_counts()

 1.0    60076
-1.0    51370
 4.0    24793
 2.0    22366
 3.0    10356
 7.0     9478
 8.0     5789
 5.0     5096
 6.0     2116
 0.0      469
Name: FES, dtype: int64

In [19]:
savepath = ('/Users/shijizhao/Documents/UCI/Courses/'
            'Intro to Data Science/Final Project/'
            'csv_hpa_2012_2016/ss16hpa_clean.csv')
df_col_row_filtered.to_csv(savepath, index=False)